extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/noop/NoOpCacheBuildRecorder.java

```diff
@@ -1,26 +1,41 @@
 package io.quarkus.cache.runtime.noop;
 
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
+import java.util.function.Supplier;
 
-import io.quarkus.arc.runtime.BeanContainer;
 import io.quarkus.cache.Cache;
+import io.quarkus.cache.CacheManager;
 import io.quarkus.cache.runtime.CacheManagerImpl;
 import io.quarkus.runtime.annotations.Recorder;
 
 @Recorder
 public class NoOpCacheBuildRecorder {
 
-    public void buildCaches(BeanContainer beanContainer, Set<String> cacheNames) {
-        // The number of caches is known at build time so we can use fixed initialCapacity and loadFactor for the caches map.
-        Map<String, Cache> caches = new HashMap<>(cacheNames.size() + 1, 1.0F);
+    public Supplier<CacheManager> getCacheManagerSupplier(Set<String> cacheNames) {
+        Objects.requireNonNull(cacheNames);
 
-        NoOpCache cache = new NoOpCache();
-        for (String cacheName : cacheNames) {
-            caches.put(cacheName, cache);
+        CacheManager cacheManager;
+        if (cacheNames.isEmpty()) {
+            cacheManager = new CacheManagerImpl(Collections.emptyMap());
+        } else {
+            // The number of caches is known at build time so we can use fixed initialCapacity and loadFactor for the caches map.
+            Map<String, Cache> caches = new HashMap<>(cacheNames.size() + 1, 1.0F);
+            NoOpCache cache = new NoOpCache();
+            for (String cacheName : cacheNames) {
+                caches.put(cacheName, cache);
+            }
+            cacheManager = new CacheManagerImpl(caches);
```

Comment:
Shouldn't all this be in the supplier's `get()`?

Reply:
No because it would delay the caches instantiation at runtime and we would no longer benefit from the `STATIC_INIT` bytecode recording.

Reply:
You sure? Because by default `SyntheticBeanBuildItem` are defined at static init.

Reply:
See https://github.com/quarkusio/quarkus/blob/master/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SyntheticBeanBuildItem.java#L75 .

Reply:
Pretty sure yes. Everything was in the supplier's `get()` method when I first added the `SyntheticBeanBuildItem`. Before pushing I wanted to verify with traces what kind of impact this would have and I saw that the caches were initialized after the JUnit tests started running. Then I moved the code out of the `get()` and everything went back to normal: initialization at build time. Maybe @mkouba could confirm the exact behavior of `SyntheticBeanBuildItem` combined with a `Supplier`.

Reply:
Are you talking build time or static init? In JVM mode, it's not the same thing.

My understanding is that static init synthetic beans will be created at static init time (and granted, not at build time).

Reply:
Sorry, my previous comment was not clear, I meant that the code inside the supplier's `get()` method didn't seem to be executed at static init time during my tests unlike the code outside of `get()`.

Here's what I see if I add [1] before the Supplier is returned (meaning outside of `get()`) and [2] at the beginning of `get()`:

```
[INFO] Running io.quarkus.cache.test.runtime.ZeroNoOpCacheTest
[1] Caches instantiation completed
2021-01-15 14:12:06,919 DEBUG [io.net.uti.int.log.InternalLoggerFactory] (Thread-905) Using SLF4J as the default logging framework
2021-01-15 14:12:07,301 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT on JVM started in 0.552s. Listening on: http://localhost:8081
2021-01-15 14:12:07,302 INFO  [io.quarkus] (main) Profile test activated.
2021-01-15 14:12:07,302 INFO  [io.quarkus] (main) Installed features: [cache, cdi]
JUnit test started
[2] Supplier's get method was called
JUnit test ended
2021-01-15 14:12:07,553 INFO  [io.quarkus] (main) Quarkus stopped in 0.019s
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.766 s - in io.quarkus.cache.test.runtime.ZeroNoOpCacheTest
```

Reply:
I should double check in debug to be sure but it will have to wait after I've finished my work day :)

Reply:
> My understanding is that static init synthetic beans will be created at static init time (and granted, not at build time).

@gsmet @gwenneg Not really. If a `SyntheticBeanBuildItem` has a `Supplier` created by a recorder, then the _real_ supplier instance is created at `STATIC INIT` or `RUNTIME INIT` (based on `SyntheticBeanBuildItem.isStaticInit()` and the `ExecutionTime` of the recorder). 

> I meant that the code inside the supplier's get() method didn't seem to be executed at static init time during my tests unlike the code outside of get().

However, the `Supplier.get()` method is invoked when the relevant bean instance is created - bean instances are created lazily. 

For example, if a build step produces a `SyntheticBeanBuildItem` which has a `Supplier` produced by a `STATIC INIT`  recorder (default) but the bean is not used during bootstrap at all, then the supplier instance is created during `STATIC INIT` but the `Supplier.get()` method can be invoked at runtime when the bean instance is needed or not at all...

Reply:
Thanks @mkouba, this confirms what I was suspecting. So unless I missed something, moving the caches instantiation inside the `get()` method would not be good for performances and the current code is fine.

Reply:
> Thanks @mkouba, this confirms what I was suspecting. So unless I missed something, moving the caches instantiation inside the `get()` method would not be good for performances and the current code is fine.

Well, you can always instantiate a bean eagerly... i.e. inside an `@Initialized(ApplicationScoped.class)` observer for static init and `StartupEvent` for runtime init [1].

[1]
https://quarkus.io/guides/cdi-reference#eager-instantiation-of-beans
https://quarkus.io/guides/lifecycle#listening-for-startup-and-shutdown-events

Reply:
Yes, but creating instances of Caffeine caches during STATIC_INIT is better for performances in native mode than doing it during an eager runtime bean instantiation, right?

Reply:
> Yes, but creating instances of Caffeine caches during `STATIC_INIT` is better for performances in native mode than doing it during an eager runtime bean instantiation, right?

You can eagerly create bean instances during `STATIC_INIT`...

Reply:
Oh sorry, your previous answer already contained everything I need. Thanks for your help @mkouba!